wordpress-mobile/WordPress-iOS
# [WIP] Make sure suggestions search doesn't block the main thread

## Issue
Improves #18818

## Description
For sites with a lot of users, the UI would potentially hitch when the user is typing an @-mention. That's because the search was done on the main thread. 

## Solution
- `showSuggestionsForWord` implementation details was refactored. The search logic is now executed within an operation queue.
- `showSuggestionsForWord` API method was changed to an async method. Which means that it no longer returns a value. Instead, callers should pass a `completionHandler`. The code that was calling this method had to be adjusted to conform to this API change.

## Test Instructions
It could be tricky to test this PR. Because you need to have access to a site that has a lot of users ( +10k ) to notice the difference. If that's the case:
1. Locate a post or comment.
2. Start a reply.
3. Insert a @-mention. 
4. You should notice that search doesn't freeze the UI.

## Regression Notes
1. Potential unintended areas of impact
Since this PR mainly refactors the internal implementation of `SuggestionsTableView`, it is possible that this refactoring could break the existing suggestions search behavior.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Unfortunately, we don't have existing automated tests for this component yet. So I relied on manually testing this PR and compared the result with the prod version.

3. What automated tests I added (or what prevented me from doing so)
`SuggestionsTableView` is currently not Unit Testable. We can open another PR to refactor this class to make it Unit Testable, then we can write the automated tests ( to double-check with my trial buddy ).

PR submission checklist:
